### ISSUE TITLE
02-data-access-layer.md - Database Management Workflow

### ISSUE THREAD

**Issue by mpskeeter** [JaimeStill/PatternsAndPractices#36]
## [Database Management Workflow](#data-access-layer)  

Whenever the entity schema is modified and you want the database to reflect the current state, a [migration](https://docs.microsoft.com/en-us/ef/core/managing-schemas/migrations/) is created and the database is updated. Given the updates to the schema made above, the following commands would be used to add a migration and update the database:  

```
{Project}.Data>dotnet ef migrations add "initial" -s ..\{Project}.Web
{Project}.Data>dotnet ef database update -s ..\{Project}.Web
```  

The `-s` flag indicates the startup project since `AppDbContext` is configured outside of the EF project. For a complete CLI reference, see the [Entity Framework Core tools reference](https://docs.microsoft.com/en-us/ef/core/miscellaneous/cli/dotnet).

---

Recommend changing:

```
{Project}.Data>dotnet ef migrations add "initial" -s ..\{Project}.Web
{Project}.Data>dotnet ef database update -s ..\{Project}.Web
```

to

```
{Project}.Data>dotnet -d ef migrations -v add "initial" -s ..\{Project}.Web
{Project}.Data>dotnet -d ef database -v update -s ..\{Project}.Web
```
---

This will provide additional outputs for debugging.

When originally issuing the command, I had previously issued:

```
dotnet run
```

then

```
PS C:\Apps\testapp\testapp.Data> dotnet -d ef migrations -v add "initial" -s ..\testapp.Web
Telemetry is: Disabled
Running C:\Program Files\dotnet\dotnet.exe "C:\Program Files\dotnet\sdk\2.2.202\DotnetTools\dotnet-ef\2.2.3\tools\netcoreapp2.2\any\dotnet-ef.dll" migrations -v add initial -s ..\testapp.Web
Process ID: 34752
Using project 'C:\Apps\testapp\testapp.Data\testapp.Data.csproj'.
Using startup project '..\testapp.Web\testapp.Web.csproj'.
Writing 'C:\Apps\testapp\testapp.Data\obj\testapp.Data.csproj.EntityFrameworkCore.targets'...
dotnet msbuild /target:GetEFProjectMetadata /property:EFProjectMetadataFile=C:\Users\Mark\AppData\Local\Temp\tmpA63D.tmp /verbosity:quiet /nologo C:\Apps\testapp\testapp.Data\testapp.Data.csproj
Telemetry is: Disabled
Writing '..\testapp.Web\obj\testapp.Web.csproj.EntityFrameworkCore.targets'...
dotnet msbuild /target:GetEFProjectMetadata /property:EFProjectMetadataFile=C:\Users\Mark\AppData\Local\Temp\tmpA8DE.tmp /verbosity:quiet /nologo ..\testapp.Web\testapp.Web.csproj
Telemetry is: Disabled
dotnet build ..\testapp.Web\testapp.Web.csproj /verbosity:quiet /nologo
Telemetry is: Disabled
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 1 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 2 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 3 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 4 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 5 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 6 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 7 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 8 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 9 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 10 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): error MSB3027: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Exceeded retry count of 10. Failed.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): error MSB3021: Unable to copy file "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process. [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]

Build FAILED.

C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 1 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 2 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 3 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 4 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 5 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 6 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 7 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 8 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 9 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): warning MSB3026: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Beginning retry 10 in 1000ms. The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): error MSB3027: Could not copy "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". Exceeded retry count of 10. Failed.  [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
C:\Program Files\dotnet\sdk\2.2.202\Microsoft.Common.CurrentVersion.targets(4374,5): error MSB3021: Unable to copy file "C:\Apps\testapp\testapp.Data\bin\Debug\netcoreapp2.2\testapp.Data.dll" to "bin\Debug\netcoreapp2.2\testapp.Data.dll". The process cannot access the file 'C:\Apps\testapp\testapp.Web\bin\Debug\netcoreapp2.2\testapp.Data.dll' because it is being used by another process. [C:\Apps\testapp\testapp.Web\testapp.Web.csproj]
    10 Warning(s)
    2 Error(s)

Time Elapsed 00:00:14.43
Microsoft.EntityFrameworkCore.Tools.CommandException: Build failed.
   at Microsoft.EntityFrameworkCore.Tools.Project.Build()
   at Microsoft.EntityFrameworkCore.Tools.RootCommand.Execute()
   at Microsoft.DotNet.Cli.CommandLine.CommandLineApplication.Execute(String[] args)
   at Microsoft.EntityFrameworkCore.Tools.Program.Main(String[] args)
Build failed.
```

By killing the 
```
dotnet run
```
command, I was able to complete the ef add command.

**Comment by JaimeStill** [comment]
I wouldn't necessarily use the `-v` command every time you update the database. I tend to only use that if there's an issue updating the database and I want to see where something is going wrong. I'll clarify this in the documentation and add an overview of the optional commands that are available to `dotnet ef migrations` and `dotnet ef database update`.

**Comment by JaimeStill** [comment]
resolved with 3360c1c8c14bcec1274c2f48c4f768641554ad23
